src/ray/util/event.h

```diff
@@ -0,0 +1,177 @@
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#ifndef RAY_EVENT_H_
+#define RAY_EVENT_H_
+
+#include <cmath>
+#include <cstring>
+#include <iomanip>
+#include <iostream>
+#include <memory>
+#include <sstream>
+#include <vector>
+#include <boost/asio/ip/host_name.hpp>
+#include <boost/filesystem.hpp>
+#include "ray/util/logging.h"
+#include "ray/util/util.h"
+#include "src/ray/protobuf/event.pb.h"
+
+namespace ray {
+
+#define RAY_EVENT(event_type, label) \
+  ::ray::RayEvent(::ray::rpc::Event_Severity::Event_Severity_##event_type, label)
+
+class LogBasedEventReporter {
+ public:
+  LogBasedEventReporter(rpc::Event_SourceType source_type, std::string log_dir);
+  
+  virtual void Init() = 0;
+
+  virtual void Report(rpc::Event &event) = 0;
+
+  virtual void Close() = 0;
+
+  virtual std::string GetReporterKey() = 0;
+
+ protected:
+  virtual std::string EventToString(rpc::Event &event);
+ 
+ protected:
+  const std::string separator_ = "|||";
+  std::string file_name_;
+  std::string log_dir_; 
+};
+
+class EventManager {
+ public:
+  static EventManager &Instance() {
+    static EventManager INSTANCE;
+    return INSTANCE;
+  }
+
+  bool IsEmpty() { return reporter_map_.empty(); }
+
+  void Publish(rpc::Event &event) {
+    for (const auto &element : reporter_map_) {
+      (element.second)->Report(event);
+    }
+  }
+
+  void AddReporter(std::shared_ptr<LogBasedEventReporter> reporter) {
+    reporter_map_.emplace(reporter->GetReporterKey(), reporter);
+  }
+
+  void ClearReporters() { reporter_map_.clear(); }
+
+ private:
+  EventManager() {}
+
+  EventManager(const EventManager &manager) = delete;
+
+  const EventManager &operator=(const EventManager &manager) = delete;
+
+ protected:
+  std::unordered_map<std::string, std::shared_ptr<LogBasedEventReporter>> reporter_map_;
+};
+
+class RayEventContext {
+ public:
+  static RayEventContext &Instance() {
+    if (instance_ == nullptr) {
+      instance_ = std::unique_ptr<RayEventContext>(new RayEventContext());
+    }
+    return *instance_;
+  }
+
+  // SetEventContext should only be called once in a process, unless ResetEventContext is
+  // called.
+  void SetEventContext(rpc::Event_SourceType source_type, std::string node_id = "",
+                       std::string job_id = "", std::string task_id = "");
+
+  void ResetEventContext();
+
+  inline void SetJobID(std::string job_id) { job_id_ = job_id; }
+
+  inline void SetNodeID(std::string node_id) { node_id_ = node_id; }
+
+  inline void SetTaskID(std::string task_id) { task_id_ = task_id; }
+
+  inline std::string GetJobID() { return job_id_; }
```

Comment:
inline const std::string& GetJobID() const { return job_id_; }

same issue in other Get/Set.